defi/crowdsourcing/src/lib.rs

```diff
@@ -0,0 +1,121 @@
+                pledged = pledged + *value;
+            }
+
+            info!("{} XRD collected from {} patrons", pledged, self.patron_entries.keys().len());
+
+            if !(Context::current_epoch() > self.last_epoch) {
+                info!("campaign ends in {} epochs", self.last_epoch - Context::current_epoch());
+            } else {
+                info!("campaign has ended. campaign holds {} unclaimed XRD.", self.collected_xrd.amount());
+            }
+        }
+
+        /*
+        Pledge XRD and become a patron.
+        */
+        pub fn pledge(&mut self, payment: Bucket) -> Bucket {
+            assert!(payment.amount() != Decimal::zero(), "you need to pay at least one XRD to become a patron.");
+            assert!(Context::current_epoch() < self.last_epoch, "campaign has already ended.");
+
+            // Create a burnable badge.
+            let patron_badge_resource = ResourceBuilder::new().metadata("name", "patron_badge").new_badge_mutable(self.patron_mint_badge.resource_def());
+            let patron_badge = self.patron_mint_badge.authorize(|mint_badge| patron_badge_resource.mint(1, mint_badge));
+
+            // Add badge and value to patron entries for this pledge.
+            self.patron_entries.insert(patron_badge.resource_address(), payment.amount());
+
+            // Put payment in collected XRD.
+            self.collected_xrd.put(payment);
+
+            // Return badge
+            patron_badge
+        }
+
+        /*
+        Recall pledge as a patron. It is allowed as long as goal hasn't been reached and last_epoch hasn't been passed.
+         */
+        pub fn recall_pledge(&mut self, patron_badge: Bucket) -> Bucket {
+            assert!(!(Context::current_epoch() > self.last_epoch && self.collected_xrd.amount() > self.goal), "campaign was successful and has ended.");
+            
+            let refund = Bucket::new(RADIX_TOKEN);
+            match self.patron_entries.get(&patron_badge.resource_address()) {
+                Some(&value) => {
+                    // Put XRD into refund bucket.
+                    refund.put(self.collected_xrd.take(value));
+                    // Remove patron entry.
+                    self.patron_entries.remove(&patron_badge.resource_address());
+                    // Authorize to burn patron badge.
+                    self.patron_mint_badge.authorize(|mint_badge| patron_badge.burn(mint_badge));
+                }
+                _ => scrypto_abort("no pledge found with provided badge")
+            }
+
+            refund
+        }
+
+        /*
+        As fundraiser, withdraw collected XRD if goal has passed, and the last_epoch has passed.
+        */
+        #[auth(fundraiser_badge_def)]
+        pub fn withdraw(&self) -> Bucket {
+            assert!(Context::current_epoch() > self.last_epoch, "campaign has not ended yet.");
+            assert!(self.collected_xrd.amount() > self.goal, "campaign did not reach it's goal.");
```

Comment:
I think you want to use ">=" instead of ">", or else you won't be able to withdraw on the end epoch

Reply:
Oh, good point! Thanks, I'm fixing in a moment.